### PR TITLE
fix: adding tests for pyclass hygiene cases

### PR DESCRIPTION
**Pull request by Zyell**
This adds the tests for pyclass hygiene mentioned in issue #4358 

This is a work in progress.  It resolves the struct case with `eq` implementation and partially resolves the enum cases as outlined in the issue.

### CODE REVIEW

File: pyo3-macros-backend/src/pyclass.rs [hunk 1]
```diff
@@ -1091,8 +1091,8 @@ fn impl_complex_enum_struct_variant_cls(
         let field_getter_impl = quote! {
             fn #field_name(slf: #pyo3_path::PyRef<Self>) -> #pyo3_path::PyResult<#field_type> {
                 match &*slf.into_super() {
-                    #enum_name::#variant_ident { #field_name, .. } => Ok(#field_name.clone()),
-                    _ => unreachable!("Wrong complex enum variant found in variant wrapper PyClass"),
+                    #enum_name::#variant_ident { #field_name, .. } => ::std::result::Result::Ok(#field_name.clone()),
```
Comment (Icxolu) [inline pyo3-macros-backend/src/pyclass.rs:1094]:
```suggestion
                    #enum_name::#variant_ident { #field_name, .. } => ::std::result::Result::Ok(::std::clone::Clone::clone(&#field_name)),
```

**Review by Icxolu — COMMENTED** [review]
Thanks for catching this! I think there are two more cases to fix here
- fully qualify the `clone` calls
- fully qualify the `From` trait in `PyClassInitializer::from`

File: pyo3-macros-backend/src/pyclass.rs (hunk 1 above)
Comment (Zyell) [reply pyo3-macros-backend/src/pyclass.rs]: Ahhh,  I see now!  I wasn't quite following the error messages before, now they make sense. :smile: .  I have pushed the fixes for from and clone.  Thanks!

**Review by Icxolu — APPROVED** [review]
Looks great, thanks again 🚀

**Pull request merged**